pkg/agent/controller/packetcapture/packetcapture_controller_linux.go

```diff
@@ -0,0 +1,699 @@
+		}
+	} else {
+		return nil, nil, fmt.Errorf("couldn't check if the file exists: %w", err)
+	}
+	writer, err := pcapgo.NewNgWriter(file, layers.LinkTypeEthernet)
+	if err != nil {
+		return nil, nil, fmt.Errorf("couldn't initialize pcap writer: %w", err)
+	}
+	return file, writer, nil
+}
+
+func (c *Controller) startPacketCapture(pc *crdv1alpha1.PacketCapture) error {
+	var err error
+	defer func() {
+		if err != nil {
+			c.cleanupPacketCapture(pc.Name)
+			err := c.updatePacketCaptureStatus(pc, crdv1alpha1.PacketCaptureFailed, fmt.Sprintf("Node: %s, Error: %+v", c.nodeConfig.Name, err), 0)
+			if err != nil {
+				klog.ErrorS(err, "failed to update PacketCapture status")
+			}
+		}
+	}()
+
+	var pod, ns string
+	if pc.Spec.Source.Pod != nil {
+		pod = pc.Spec.Source.Pod.Name
+		ns = pc.Spec.Source.Pod.Namespace
+	} else {
+		pod = pc.Spec.Destination.Pod.Name
+		ns = pc.Spec.Destination.Pod.Namespace
+	}
+	pcState := &packetCaptureState{name: pc.Name}
+	podInterfaces := c.interfaceStore.GetContainerInterfacesByPod(pod, ns)
+	if len(podInterfaces) == 0 {
+		return nil
+	}
+	device := podInterfaces[0].InterfaceName
+	var matchPacket *binding.Packet
+	matchPacket, err = c.createMatchPacket(pc)
+	if err != nil {
+		return err
+	}
+	klog.V(2).InfoS("PacketCapture trying to match packet", "name", pc.Name, "packet", *matchPacket)
+	pcState.maxNumCapturedPackets = pc.Spec.CaptureConfig.FirstN.Number
+	file, writer, err := getPacketFileAndWriter(pc.Name)
+	if err != nil {
+		return err
+	}
+	pcState.pcapngFile = file
+	pcState.pcapngWriter = writer
+	pcState.updateRateLimiter = rate.NewLimiter(rate.Every(captureStatusUpdatePeriod), 1)
+	timeout := crdv1alpha1.DefaultPacketCaptureTimeout
+	if pc.Spec.Timeout != nil {
+		timeout = *pc.Spec.Timeout
+	}
+	exp := genBPFFilterStr(matchPacket, pc.Spec.Packet)
+	klog.V(2).InfoS("Starting capture on device", "name", pc.Name, "device", device, "filter", exp)
+	err = c.performCapture(pcState, device, exp, time.Duration(timeout)*time.Second)
+	return err
+}
+
+// genBPFFilterStr generate BPF filter string based on the origin PacketCapture spec and
+// parsed ip / port info (matchPacket) from it. An example generated filter string would be like:
+// 'src 192.168.0.1 and dst 192.168.0.2 and src port 8080 and dst port 8081 and tcp and tcp[tcpflags] & (tcp-syn|tcp-ack) != 0'
+func genBPFFilterStr(matchPacket *binding.Packet, packetSpec *crdv1alpha1.Packet) string {
+	exp := ""
+	protocol := packetSpec.Protocol
+	if protocol != nil {
+		if protocol.Type == intstr.Int {
+			if val, ok := protocolMap[protocol.IntValue()]; ok {
+				exp += val
+			} else {
+				// go-pcap didn't support proto number for now.
+				exp += "proto " + strconv.Itoa(protocol.IntValue())
+			}
+		} else {
+			exp += strings.ToLower(protocol.String())
+		}
+	}
+	if exp != "" {
+		if exp == "icmp" || exp == "icmp6" {
+			// go-pcap bug, see:https://github.com/packetcap/go-pcap/issues/59
+			// cannot use `and` now
+			exp += " "
```

Comment:
Maybe add a TODO to fix this after the go-pcap bug is fixed.

Reply:
Hi, we have choose a solution to create our own bpf instructions parser, go-pcap is not needed anymore. You can checkout the latest changes.